src/Npgsql/Internal/TypeHandlers/NumericHandlers/NumericHandler.cs

```diff
@@ -103,30 +124,116 @@ public override decimal Read(NpgsqlReadBuffer buf, int len, FieldDescription? fi
+                    SignPinf => throw new InvalidCastException("Numeric Infinity not supported by BigInteger"),
+                    SignNinf => throw new InvalidCastException("Numeric -Infinity not supported by BigInteger"),
+                    _ => throw new InvalidCastException("Numeric special value not supported"),
+                };
+
+            if (weightRight < 0)
+            {
+                await buf.Skip(groups * sizeof(ushort), async);
+                throw new InvalidCastException("Numeric value with non-zero fractional digits not supported by BigInteger");
+            }
+
+            var digits = new ushort[groups];
+
+            for (var i = 0; i < groups; i++)
+            {
+                await buf.Ensure(sizeof(ushort), async);
+                digits[i] = buf.ReadUInt16();
+            }
+
+            // Calculate powers 10^8, 10^16, 10^32, ...
+            // We should have the last calculated power to be less than the input
+            var lenPow = 2; // 2 ushorts fit in one uint, represents 10^8
+            var numPowers = 0;
+            while (lenPow < weightLeft + 1)
+            {
+                lenPow <<= 1;
+                ++numPowers;
+            }
+            var factors = numPowers > 0 ? new BigInteger[numPowers] : null;
+            if (numPowers > 0)
+            {
+                factors![0] = new BigInteger(100000000U);
+                for (var i = 1; i < numPowers; i++)
+                    factors[i] = factors[i - 1] * factors[i - 1];
+            }
+
+            var result = ToBigIntegerInner(0, weightLeft + 1, digits, factors);
+            return sign == SignPositive ? result : -result;
+
+            static BigInteger ToBigIntegerInner(int offset, int length, ushort[] digits, BigInteger[]? factors)
+            {
+                if (length <= 2)
+                {
+                    var r = 0U;
+                    for (var i = offset; i < offset + length; i++)
+                    {
+                        r *= 10000U;
+                        r += i < digits.Length ? digits[i] : 0U;
+                    }
+                    return r;
+                }
+                else
+                {
+                    // Split the input into two halves, the lower one should be a power of two in digit length,
+                    // then multiply the higher part with a precomputed power of 10^8 and add the results.
+                    var lenFirstHalf = 2 << 1; // 2 ushorts fit in one uint, skip 1 since we've already covered the base case.
+                    var pos = 0;
+                    while (lenFirstHalf < length)
+                    {
+                        lenFirstHalf <<= 1;
+                        ++pos;
+                    }
+                    var factor = factors![pos];
+                    lenFirstHalf >>= 1;
+                    var lo = ToBigIntegerInner(offset + length - lenFirstHalf, lenFirstHalf, digits, factors);
```

Comment:
Tell that to the one who implemented this method https://github.com/dotnet/runtime/blob/0c793920f62dd2711e6423a3425110e5cb7acfde/src/libraries/System.Runtime.Numerics/src/System/Numerics/BigNumber.cs#L441 which is extremely slow and creates huge amounts of allocations ;)

Reply:
Yeah, seems like this type was passed over in the general .NET Core optimization push... :/

Up to you if you want to invest in this or not :) I'm fine either way.

Reply:
You mean send a PR to dotnet runtime repo?

To answer your previous question, it's not that simple to avoid the allocations for temporary results. The simplest way seems to be to create a replica of BigInteger and its arithmetics that uses a reusable buffer pool, but I think that's not a good way to go for a db driver just to optimize away some allocations...

Reply:
I meant in Npgsql, unless you feel like attacking the type itself :)

In any case, definitely no issues from me on this - was just pointing it out in case you missed it.